build/DownloadCFamilyPlugin/DownloadCFamilyPlugin.csproj

```diff
@@ -69,4 +42,96 @@
     
   </Target>
 
+
+  <!-- ******************************************** -->
+  <!-- TypeScript jar handling -->
+  <!-- ******************************************** -->
+  <UsingTask TaskName="DownloadAndExtractSonarJS" AssemblyFile="$(OutputPath)DownloadCFamilyPlugin.dll" />
+
+  <Target Name="EmbedTypeScriptFiles" DependsOnTargets="ValidateInputs" >
+    <Message Importance="high" Text="Fetching TypeScript files..." />
+
+    <DownloadAndExtractSonarJS DownloadUrl="$(PluginUrl)">
+      <Output TaskParameter="FilesToEmbed" ItemName="FilesToEmbed" />
+      <Output TaskParameter="PackageDirectoryToEmbed" PropertyName="PackageDirectoryToEmbed" />
+    </DownloadAndExtractSonarJS>
```

Comment:
The process for dealing with SonarJS outputs is slightly more complicated than for the CFamily plugin. 

For the CFamily, we can copy everything in to a single flat directory.
For SonarJS, we want to preserve the directory structure of the `packages` folder.

The way the MSBuild `Copy` task works means we need to handle the folder-copying separately, so this task returns two outputs - a list of "flat" files, and the "packages" directory.

Overall, I'm not convinced about using tasks and targets for this - it might be simpler just to write an exe and execute it directly. However, it was easier to following the existing approach. It's internal build tooling, so the main thing for now is that it produces the correct output.

Reply:
I agree, in typescript's case it feels a bit over-complicated for what we're trying to achieve. We can revisit this in the future. I just hope it never breaks when you're on a holiday :D